sklearn/neighbors/tests/test_lmnn.py

```diff
@@ -0,0 +1,677 @@
+    params = {'n_neighbors': 3, 'max_impostors': 5, 'random_state': 1,
+              'max_iter': 10, 'init': 'identity'}
+
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_1 = lmnn.components_
+
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_2 = lmnn.components_
+
+    # This assertion fails on 32bit systems if init='pca'
+    assert_allclose(transformation_1, transformation_2)
+
+    params['random_state'] = 2
+    lmnn = LargeMarginNearestNeighbor(**params)
+    lmnn.fit(X, y)
+    transformation_3 = lmnn.components_
+
+    assert not np.allclose(transformation_2, transformation_3)
+
+
+def test_same_lmnn_parallel():
+    X, y = datasets.make_classification(n_samples=30, n_features=5,
+                                        n_redundant=0, random_state=0)
+    X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3)
+    lmnn.fit(X_train, y_train)
+    components = lmnn.components_
+
+    lmnn.set_params(n_jobs=3)
+    lmnn.fit(X_train, y_train)
+    components_parallel = lmnn.components_
+
+    assert_array_almost_equal(components, components_parallel)
+
+
+def test_singleton_class():
+    X = iris_data
+    y = iris_target
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.3, stratify=y)
+
+    # one singleton class
+    singleton_class = 1
+    ind_singleton, = np.where(y_tr == singleton_class)
+    y_tr[ind_singleton] = 2
+    y_tr[ind_singleton[0]] = singleton_class
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, max_iter=30)
+    lmnn.fit(X_tr, y_tr)
+
+    # One non-singleton class
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.3, stratify=y)
+    ind_1, = np.where(y_tr == 1)
+    ind_2, = np.where(y_tr == 2)
+    y_tr[ind_1] = 0
+    y_tr[ind_1[0]] = 1
+    y_tr[ind_2] = 0
+    y_tr[ind_2[0]] = 2
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, max_iter=30)
+    assert_raise_message(ValueError,
+                         'LargeMarginNearestNeighbor needs at least 2 '
+                         'non-singleton classes, got 1.',
+                         lmnn.fit, X_tr, y_tr)
+
+
+def test_convergence_warning():
+
+    lmnn = LargeMarginNearestNeighbor(n_neighbors=3, max_iter=2, verbose=1)
+    cls_name = lmnn.__class__.__name__
+    assert_warns_message(ConvergenceWarning,
+                         '[{}] LMNN did not converge'.format(cls_name),
+                         lmnn.fit, iris_data, iris_target)
+
+
+def test_paired_distances_blockwise():
+    n, d = 10000, 100  # 4 or 8 MiB
+    X = rng.rand(n, d)
+    ind_a = rng.permutation(n)
+    ind_b = rng.permutation(n)
+
+    distances = paired_euclidean_distances(X[ind_a], X[ind_b])
+    distances_blockwise = _paired_distances_blockwise(
+        X, ind_a, ind_b, squared=False, block_size=1)
+    assert_array_equal(distances, distances_blockwise)
+
+
+def test_euclidean_distances_without_checks():
```

Comment:
since `_euclidean_distances_without_checks` is in `sklearn.utils.extmath`, maybe the test should be in `sklearn.utils.tests` ?

Reply:
Yeap. Done.